pkg/datapath/linux/node.go

```diff
@@ -0,0 +1,530 @@
+	n.mutex.Lock()
+	defer n.mutex.Unlock()
+
+	n.nodes[newNode.Identity()] = &newNode
+
+	if n.isInitialized {
+		return n.nodeUpdate(&oldNode, &newNode)
+	}
+
+	return nil
+}
+
+func (n *linuxNodeHandler) nodeUpdate(oldNode, newNode *node.Node) error {
+	var (
+		oldIP4Cidr, oldIP6Cidr *net.IPNet
+		oldIP4, oldIP6         net.IP
+		newIP4                 = newNode.GetNodeIP(false)
+		newIP6                 = newNode.GetNodeIP(true)
+		firstAddition          = oldNode == nil
+	)
+
+	if oldNode != nil {
+		oldIP4Cidr = oldNode.IPv4AllocCIDR
+		oldIP6Cidr = oldNode.IPv6AllocCIDR
+		oldIP4 = oldNode.GetNodeIP(false)
+		oldIP6 = oldNode.GetNodeIP(true)
+	}
+
+	if newNode.IsLocal() {
+		if !n.nodeConfig.DisableLocalNodeRoute {
+			n.updateOrRemoveNodeRoutes([]*net.IPNet{oldIP4Cidr}, []*net.IPNet{newNode.IPv4AllocCIDR})
+			n.updateOrRemoveNodeRoutes([]*net.IPNet{oldIP6Cidr}, []*net.IPNet{newNode.IPv6AllocCIDR})
+		}
+		return nil
+	}
+
+	if n.nodeConfig.EnableAutoDirectRouting {
+		n.updateDirectRoute(oldIP4Cidr, newNode.IPv4AllocCIDR, oldIP4, newIP4, firstAddition, n.nodeConfig.EnableIPv4)
+		n.updateDirectRoute(oldIP6Cidr, newNode.IPv6AllocCIDR, oldIP6, newIP6, firstAddition, n.nodeConfig.EnableIPv6)
+		return nil
+	} else if oldNode == nil {
+		// When direct routing is disabled, then the initial node
+		// addition triggers a removal of eventual old tunnel map
+		// entries.
+		n.deleteDirectRoute(newNode.IPv4AllocCIDR, newIP4)
+		n.deleteDirectRoute(newNode.IPv6AllocCIDR, newIP6)
+	}
+
+	if n.nodeConfig.EnableEncapsulation {
+		// Update the tunnel mapping of the node. In case the
+		// node has changed its CIDR range, a new entry in the
+		// map is created and the old entry is removed.
+		updateTunnelMapping(oldIP4Cidr, newNode.IPv4AllocCIDR, oldIP4, newIP4, firstAddition, n.nodeConfig.EnableIPv4)
+		// Not a typo, the IPv4 host IP is used to build the IPv6 overlay
+		updateTunnelMapping(oldIP6Cidr, newNode.IPv6AllocCIDR, oldIP4, newIP4, firstAddition, n.nodeConfig.EnableIPv6)
+
+		if !n.nodeConfig.UseSingleClusterRoute {
+			n.updateOrRemoveNodeRoutes([]*net.IPNet{oldIP4Cidr}, []*net.IPNet{newNode.IPv4AllocCIDR})
+			n.updateOrRemoveNodeRoutes([]*net.IPNet{oldIP6Cidr}, []*net.IPNet{newNode.IPv6AllocCIDR})
+		}
+
+		return nil
+	} else if oldNode == nil {
+		// When encapsulation is disabled, then the initial node addition
+		// triggers a removal of eventual old tunnel map entries.
+		deleteTunnelMapping(newNode.IPv4AllocCIDR, true)
+		deleteTunnelMapping(newNode.IPv6AllocCIDR, true)
+
+		if rt, _ := n.lookupNodeRoute(newNode.IPv4AllocCIDR); rt != nil {
+			n.deleteNodeRoute(newNode.IPv4AllocCIDR)
+		}
+		if rt, _ := n.lookupNodeRoute(newNode.IPv6AllocCIDR); rt != nil {
+			n.deleteNodeRoute(newNode.IPv6AllocCIDR)
+		}
+	}
+
+	return nil
+}
+
+func (n *linuxNodeHandler) NodeDelete(oldNode node.Node) error {
+	n.mutex.Lock()
+	defer n.mutex.Unlock()
+
+	delete(n.nodes, oldNode.Identity())
+
+	if n.isInitialized {
+		return n.nodeDelete(&oldNode)
```

Comment:
Small optimization, this line should only be called if `n.nodes` contained the `oldNode` in the map